packages/@aws-cdk/aws-iam/lib/private/postprocess-policy-document.ts

```diff
@@ -0,0 +1,149 @@
+import * as cdk from '@aws-cdk/core';
+import { LITERAL_STRING_KEY } from '../util';
+import { mergeStatements } from './merge-statements';
+
+/**
+ * A Token postprocesser for policy documents
+ *
+ * Removes duplicate statements, merges statements, and assign Sids if necessary
+ *
+ * Because policy documents can contain all kinds of crazy things,
+ * we do all the necessary work here after the document has been mostly resolved
+ * into a predictable CloudFormation form.
+ */
+export class PostProcessPolicyDocument implements cdk.IPostProcessor {
+  constructor(private readonly autoAssignSids: boolean, private readonly minimize: boolean) {
+  }
+
+  public postProcess(input: any, _context: cdk.IResolveContext): any {
+    if (!input || !input.Statement) {
+      return input;
+    }
+
+    if (this.minimize) {
+      input.Statement = mergeStatements(input.Statement);
+    }
+
+    // Also remove full-on duplicates (this will not be necessary if
+    // we minimized, but it might still dedupe statements we didn't
+    // minimize like 'Deny' statements, and definitely is still necessary
+    // if we didn't minimize)
+    const jsonStatements = new Set<string>();
+    const uniqueStatements: any[] = [];
+
+    for (const statement of input.Statement) {
+      const jsonStatement = JSON.stringify(statement);
+      if (!jsonStatements.has(jsonStatement)) {
+        uniqueStatements.push(statement);
+        jsonStatements.add(jsonStatement);
+      }
+    }
+
+    // assign unique SIDs (the statement index) if `autoAssignSids` is enabled
+    const statements = uniqueStatements.map((s, i) => {
+      if (this.autoAssignSids && !s.Sid) {
+        s.Sid = i.toString();
+      }
+
+      return s;
+    });
+
+    return {
+      ...input,
+      Statement: statements,
+    };
+  }
+}
+
+// An IAM value is a string or a CloudFormation intrinsic
+export type IamValue = string | {} | Array<string | {}>;
+
+export interface StatementSchema {
+  readonly Sid?: string;
+  readonly Effect?: string;
+  readonly Principal?: Record<string, IamValue>;
+  readonly NotPrincipal?: Record<string, IamValue>;
+  readonly Resource?: IamValue;
+  readonly NotResource?: IamValue;
+  readonly Action?: IamValue;
+  readonly NotAction?: IamValue;
+  readonly Condition?: any;
+}
+
+
+export function normalizeStatement(s: StatementSchema) {
+  return noUndef({
+    Action: _norm(s.Action, { unique: true }),
+    NotAction: _norm(s.NotAction, { unique: true }),
+    Condition: _norm(s.Condition),
+    Effect: _norm(s.Effect),
+    Principal: _normPrincipal(s.Principal),
+    NotPrincipal: _normPrincipal(s.NotPrincipal),
+    Resource: _norm(s.Resource, { unique: true }),
+    NotResource: _norm(s.NotResource, { unique: true }),
+    Sid: _norm(s.Sid),
+  });
+
+  function _norm(values: any, { unique }: { unique: boolean } = { unique: false }) {
+
+    if (typeof(values) === 'undefined') {
+      return undefined;
+    }
```

Comment:
This is a more compact, slightly more efficient way to express the same intent (although it also captures `null` which is so easily forgotten downstream)

```suggestion
    if (values == null) {
      return undefined;
    }
```